hail/python/test/hailtop/aiotools/test_copy.py

```diff
@@ -0,0 +1,351 @@
+    dest_base = await fresh_dir(fs, bases, dest_scheme)
+
+    # make sure dest_base exists
+    async with await fs.create(f'{dest_base}keep'):
+        pass
+
+    yield fs, src_base, dest_base
+
+
+@pytest.mark.asyncio
+async def test_copy_behavior(copy_test_context, test_spec):
+    fs, src_base, dest_base = copy_test_context
+
+    result = await run_test_spec(fs, test_spec, src_base, dest_base)
+    try:
+        expected = test_spec['result']
+
+        dest_scheme = url_scheme(dest_base)
+        if (dest_scheme == 'gs'
+                and 'files' in result
+                and expected.get('exception') in ('IsADirectoryError', 'NotADirectoryError')):
+            return
+
+        assert result == expected, (test_spec, result, expected)
+    except Exception:
+        print(test_spec)
+        raise
+
+
+async def expect_file(fs, path, expected):
+    async with await fs.open(path) as f:
+        actual = (await f.read()).decode('utf-8')
+    assert actual == expected, (actual, expected)
+
+
+class DidNotRaiseError(Exception):
+    pass
+
+
+class RaisedWrongExceptionError(Exception):
+    pass
+
+
+class RaisesOrGS:
+    def __init__(self, dest_base, expected_type):
+        self._gs = url_scheme(dest_base) == 'gs'
+        self._expected_type = expected_type
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        # gs can succeed or throw
+        if type is None:
+            if not self._gs:
+                raise DidNotRaiseError()
+        elif type != self._expected_type:
+            raise RaisedWrongExceptionError(type)
+
+        # suppress exception
+        return True
+
+@pytest.mark.asyncio
+async def test_copy_doesnt_exist(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    with pytest.raises(FileNotFoundError):
+        await fs.copy(Transfer(f'{src_base}a', dest_base))
+
+
+@pytest.mark.asyncio
+async def test_copy_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base.rstrip('/')))
+
+    await expect_file(fs, f'{dest_base}a', 'src/a')
+
+
+@pytest.mark.asyncio
+async def test_copy_large_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    # mainly needs to be larger than the transfer block size (8K)
+    contents = secrets.token_bytes(1_000_000)
+    async with await fs.create(f'{src_base}a') as f:
+        await f.write(contents)
+
+    await fs.copy(Transfer(f'{src_base}a', dest_base.rstrip('/')))
+
+    async with await fs.open(f'{dest_base}a') as f:
+        copy_contents = await f.read()
+    assert copy_contents == contents
+
+
+@pytest.mark.asyncio
+async def test_copy_rename_file(copy_test_context):
+    fs, src_base, dest_base = copy_test_context
+
+    await create_test_file(fs, 'src', src_base, 'a')
+
+    await fs.copy(Transfer(f'{src_base}a', f'{dest_base}x'))
```

Comment:
What happens if this is `Target.TARGET_DIRECTORY` and `x` doesn't exist?

Reply:
OK, this is an interesting example and it is possible we want to change the behavior.  I added a test test_copy_file_dest_target_directory_doesnt_exist to test this.  It works, in which I mean it creates dest/x/a on both the local filesystem and gs.  This is because (1) if you set treat_dest_as=Transfer.TARGET_DIR, it doesn't actually stat the destination and just treats it as a directory.  Then when we go to copy src/a to dest/x/a, SourceCopier._copy_file creates the destination if it doesn't exist.  This is because when you copy cp dir dest and dir/a/b/c exists, we need to create dest/a/b.  In this case, dest itself doesn't exist, so we create it.  I'm not unhappy with the behavior, but it may be surprising since cp throws an error.

Either way, it won't throw an error on gs.

Options going forward:
 - Leave behavior as-is.
 - Always stat dest and give an error if it doesn't exist.  This will do unnecessary work for errors, which I've avoided so far.
 - Don't create the dest itself in _copy_file.  Then we'd get a FileNotFoundError on file, and no error on gs.